File: android/app/src/main/java/com/bluelink/android/session/TransferPauseController.kt
package com.bluelink.android.session

import com.bluelink.android.domain.TransferItem
import com.bluelink.android.domain.TransferStatus
import kotlinx.coroutines.CompletableDeferred

/** Local and remote pause requests are independent; neither side can clear the other's pause. */
internal class TransferPauseController(private val notify: (TransferItem) -> Unit) {
    private val lock = Any()
    private var localPaused = false
    private var remotePaused = false
    private var failure: Throwable? = null
    private var resumed = CompletableDeferred<Unit>().also { it.complete(Unit) }
    @Volatile var item: TransferItem? = null
        private set

    fun report(update: TransferItem) = synchronized(lock) {
        // A late block acknowledgement or canceled worker must not replace a final result.
        if (item?.status in setOf(TransferStatus.COMPLETED, TransferStatus.CANCELED, TransferStatus.FAILED, TransferStatus.REJECTED)) return@synchronized
        if (item?.status?.let { it !in pausable } == true && update.status in pausable) return@synchronized
        publish(update)
    }

    fun setPaused(local: Boolean, paused: Boolean): Boolean = synchronized(lock) {
        val current = item ?: return@synchronized false
        if (failure != null || current.status !in pausable) return@synchronized false
        if ((if (local) localPaused else remotePaused) == paused) return@synchronized false
        if (local) localPaused = paused else remotePaused = paused
        if (localPaused || remotePaused) {
            if (resumed.isCompleted) resumed = CompletableDeferred()
        } else resumed.complete(Unit)
        publish(current.copy(status = TransferStatus.RESUMING))
        true
    }

    private fun publish(update: TransferItem) {
        val status = if (update.status !in pausable) update.status else when {
            localPaused -> TransferStatus.PAUSED
            remotePaused -> TransferStatus.REMOTE_PAUSED
            else -> update.status
        }
        val projected = update.copy(status = status)
        item = projected
        notify(projected)
    }

    suspend fun awaitResumed() {
        while (true) {
            val signal = synchronized(lock) {
                failure?.let { throw it }
                if (!localPaused && !remotePaused) return
                resumed
            }
            signal.await()
        }
    }

    fun fail(cause: Throwable) = synchronized(lock) {
        failure = cause
        resumed.completeExceptionally(cause)
    }

    private companion object {
        val pausable = setOf(TransferStatus.OFFERED, TransferStatus.QUEUED, TransferStatus.TRANSFERRING,
            TransferStatus.PAUSED, TransferStatus.REMOTE_PAUSED, TransferStatus.RESUMING)
    }
}

File: windows/BlueLink.App/Session/TransferPauseController.cs
using BlueLink.Domain;

namespace BlueLink.Session;

/// <summary>Local and remote pause requests remain independent while late block acknowledgements arrive.</summary>
internal sealed class TransferPauseController(TransferItem item, Action<TransferItem> notify)
{
    private readonly object _gate = new();
    private bool _localPaused, _remotePaused;
    private Exception? _failure;
    private TaskCompletionSource _resumed = Ready();
    public TransferItem Item { get; } = item;
    public bool SetPaused(bool local, bool paused)
    {
        lock (_gate)
        {
            if (_failure is not null || !Pausable(Item.Status)) return false;
            if ((local ? _localPaused : _remotePaused) == paused) return false;
            if (local) _localPaused = paused; else _remotePaused = paused;
            if (_localPaused || _remotePaused)
            {
                if (_resumed.Task.IsCompleted) _resumed = new(TaskCreationOptions.RunContinuationsAsynchronously);
            }
            else _resumed.TrySetResult();
            Publish(TransferStatus.Resuming);
            return true;
        }
    }
    public void Report(TransferStatus status, long? completedBytes = null)
    {
        lock (_gate)
        {
            if (Item.Status is TransferStatus.Completed or TransferStatus.Canceled or TransferStatus.Failed or TransferStatus.Rejected) return;
            if (!Pausable(Item.Status) && Pausable(status)) return;
            if (completedBytes is { } bytes) Item.CompletedBytes = bytes;
            Publish(status);
        }
    }
    private void Publish(TransferStatus status)
    {
        Item.Status = !Pausable(status) ? status : _localPaused ? TransferStatus.Paused :
            _remotePaused ? TransferStatus.RemotePaused : status;
        notify(Item);
    }
    public async Task WaitAsync(CancellationToken token)
    {
        while (true)
        {
            Task signal;
            lock (_gate)
            {
                if (_failure is { } failure) throw failure;
                if (!_localPaused && !_remotePaused) return;
                signal = _resumed.Task;
            }
            await signal.WaitAsync(token).ConfigureAwait(false);
        }
    }
    public void Fail(Exception failure)
    {
        lock (_gate) { _failure = failure; _resumed.TrySetException(failure); }
    }
    private static bool Pausable(TransferStatus status) => status is TransferStatus.Offered or TransferStatus.Queued or
        TransferStatus.Transferring or TransferStatus.Paused or TransferStatus.RemotePaused or TransferStatus.Resuming;
    private static TaskCompletionSource Ready()
    {
        var signal = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
        signal.SetResult(); return signal;
    }
}
